queue: k_queue_get: Fix NULL return

Vudentz opened the pull request:
k_queue_get shall never return NULL when timeout is K_FOREVER which can
happen when a higher priority thread cancel/take an item before the
waiting thread.

Fixes  #4358

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
queue: k_queue_get: Fix NULL return

k_queue_get shall never return NULL when timeout is K_FOREVER which can
happen when a higher priority thread cancel/take an item before the
waiting thread.

Fixes issue #4358

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on kernel/queue.c:
```diff
@@ -207,19 +207,23 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
-		return NULL;
-	}
+	do {
+		err = k_poll(&event, 1, timeout);
```
I think you need to move the "event.state = K_POLL_STATE_NOT_READY;" into the loop so that it gets set every time before you call k_poll. That's at least how Ben instructed me to use the API when I originally took it into use in our Bluetooth code (which also calls it in a loop).

carlescufi replied:
Agreed, as stated in here: https://www.zephyrproject.org/doc/kernel/other/polling.html#using-k-poll

cvinayak commented on kernel/queue.c:
```diff
@@ -207,19 +207,23 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
+		key = irq_lock();
+		val = sys_slist_get(&queue->data_q);
+		irq_unlock(key);
+	} while(!val && timeout == K_FOREVER);
```
Prefer a onetime if-clause to check timeout == K_FOREVER, avoids extra branch-instructions in a loop.

cvinayak commented on kernel/queue.c:
```diff
@@ -207,19 +207,23 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
+		 * irq_lock.
+		 */
+		key = irq_lock();
+		val = sys_slist_get(&queue->data_q);
```
if canceled, val will be NULL, does this satisfy the cancelled usecase?

carlescufi replied:
If cancelled won't k_poll() return -EINT?

Vudentz replied:
k_poll doesn't return -EINT, not sure if there is anything that returns EINT actually.

pfalcon replied:
Indeed, not sure how this would work with k_fifo_cancel_wait().

carlescufi replied:
So what do we expect to happen when k_fifo_cancel_wait() is called? we expect k_poll() to return err != 0?

pfalcon replied:
Well, we expect k_fifo_get() to return with NULL. How that is achieved is an implementation detail, but that case isn't handled by this patch.

Vudentz replied:
It is clearly not handling it, weird I thought we had a test for it, apparently not. It is probably because only the fifo tests are testing this which does not select CONFIG_POLL.

cvinayak commented on kernel/queue.c:
```diff
@@ -205,21 +205,25 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
+		key = irq_lock();
+		val = sys_slist_get(&queue->data_q);
+		irq_unlock(key);
+	} while (!val && timeout == K_FOREVER);
```
Can we have a if-clause outside loop to check timeout == K_FOREVER? This is to avoid branching in loop. Lesser CPU time.

Vudentz replied:
Hmm where would you put it though? Btw, are you sure this would be optimized away by the compiler?

carlescufi replied:
@Vudentz I think @cvinayak means an early check to iterate to avoid having to test for K_FOREVER in every iteration of the loop

cvinayak replied:
@Vudentz my suggested diff to current code in this PR: https://gist.github.com/cvinayak/9bb2e7dc5da943df926cf48c4df3d169

or plain code:
https://gist.github.com/cvinayak/5f5dfed8b018038f98c34e7f6b2f68c8

Vudentz replied:
@cvinayak Since the loop does actually contain a call that can sleep/pend I don't think we need such optimization since it is very unlikely we will get many NULLs in a row to cause a substantial difference.

cvinayak replied:
@Vudentz its hard for me to see a while-condition with a variable that does not change inside the loop. There is always the possibility with the correct frequency of IRQ this loop can run for many iterations.

andyross replied:
FWIW: the compiler will surely hoist that "timeout == K_FOREVER" out of the loop.  I guess we could check the assembly to be sure, but I don't see anything with this as written.  It's clear and relying on a compiler optimizer is something we always do anyway.

Vudentz committed:
poll: k_poll: Return -EINTR if not ready

In case _handle_obj_poll_events is called with K_POLL_STATE_NOT_READY
set -EINTR as return to the poller thread.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
queue: k_queue_cancel_wait: Fix not interrupting other threads

When k_poll is being used k_queue_cancel_wait shall mark the state as
K_POLL_STATE_NOT_READY so other threads will get properly notified with
a NULL pointer return.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

cvinayak commented on kernel/queue.c:
```diff
@@ -205,21 +203,25 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
-		return NULL;
-	}
+		err = k_poll(&event, 1, timeout);
+		if (err) {
```
Are you missing the check for -EINVAL?

carlescufi replied:
Does it matter? if it's any error you return NULL anyway?

cvinayak replied:
ah ok.

Vudentz replied:
Yep, any error here will cause it to return NULL which is what the user of k_queue_cancel_wait would expect according to its documentation.

Vudentz committed:
tests: fifo: Add prj_poll.conf

This enables testing k_fifo test with CONFIG_POLL=y.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi reviewed (approved):
This patch guarantees a non-NULL return value for k_queue_get(). Verified with testbench

carlescufi commented:
@nashif @andrewboie @andyross this re-establishes the "queue contract" between kernel and application, returning always != NULL. We'd like to backport this to 1.9.2 in order to provide a stable 1.9.x build. Any comments appreciated.

pfalcon commented on kernel/poll.c:
```diff
@@ -292,7 +292,8 @@ static int _signal_poll_event(struct k_poll_event *event, u32_t state,
 	_abort_thread_timeout(thread);
-	_set_thread_return_value(thread, 0);
+	_set_thread_return_value(thread,
+				 state == K_POLL_STATE_NOT_READY ? -EINVAL : 0);
```
Is EINVAL a good error here? It's meaning is "invalid *input* value" or "invalid value", there's nothing invalid in K_POLL_STATE_NOT_READY, it's just a special value to act upon. Previously, EINTR was mentioned, or we definitely use EAGAIN in somewhat similar contexts.

cvinayak replied:
Would -ECANCELED be more relevant?

Vudentz replied:
Changed that to -EINTR, for whatever reason I thought that was not available on zephyr.

pfalcon commented:
Tested to not affect samples/net/sockets/dumb_http_server. (I'd like to do more testing with MicroPython before posting "approve" from my side.)

carlescufi commented:
Retested on top of all 1.9.2 patches, no issues.

pfalcon reviewed (approved):
In addition to my previous comment, also tested with MicroPython networking sample apps: test_dl - works as expected; http_server - does have some issues (unlike the C code of dump_http_server), but they also happen without this patch, and I didn't notice any correlation between nature/frequency of the issues and presence/absence on this patch.

So, clear from my side.

andyross commented on kernel/queue.c:
```diff
@@ -205,21 +205,25 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
+				K_POLL_STATE_FIFO_DATA_AVAILABLE);
+
+		/* sys_slist_* aren't threadsafe, so must be always protected by
+		 * irq_lock.
```
Not sure you really need this comment.  It's clear what is happening from the context.

pfalcon replied:
That comment was added by me in the previous commit (this patch just moves it around). I doubt that it's "clear what is happening", because nobody was able to tell what's wrong in the original issue https://github.com/zephyrproject-rtos/zephyr/issues/4022 for 1.5 months.

andyross reviewed (changes requested):
Can you explain why exactly this is happening?  This sounds like a bug to me.  It's true that in classic Unix a "system call interrupted" condition would produce behavior like this, but that's not the way k_poll() works.

And in any case, why wouldn't -EAGAIN be a better return value?

Basically EINTR is old-school unix voodoo, not something with a clear meaning to a modern system.

andyross reviewed (approved):
OK, never mind.  I see where that EINTR is coming from in the future patch.  Honestly I think think you want EAGAIN here ("another thread cancelled it" is not what Unix meant by EINTR), but that's minor.

pfalcon commented:
>  Honestly I think think you want EAGAIN here ("another thread cancelled it" is not what Unix meant by EINTR), but that's minor.

I'd +0.5 that. (I was just happy it didn't stay EINVAL as it was originally, but indeed, EINTR may spur some wrong associations in someone.)
